torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py

```diff
@@ -65,6 +65,51 @@ def decompress(fut):
     return fut.then(decompress)
 
 
+class OptimizerHookState(object):
+    """
+    Holds state for running optimizer in-line after DDP communication hook.
+    Currently contains optimizer class as well as optimizer args.
+    """
+    __slots__ = ['functional_optim_cls', 'functional_optim_args', 'functional_optimizer']
+
+    def __init__(self, functional_optim_cls, *functional_optim_args):
+        # TODO: Support kwargs
+        self.functional_optim_cls = functional_optim_cls
+        self.functional_optim_args = functional_optim_args
```

Comment:
Is it possible that we may have tensor args here? If so, shall we co-locate them on the same device?

Reply:
@SciPioneer for SGD AFAIK the arguments are all non-tensor, but future optimizers may have tensor arguments. What does colocating the tensors on the same device (as opposed to keeping them on device user specified) allow?